File: strategy.go
package main

import (
	p "apivapt/parsers"
	"net/http"
)

type ParserContext struct {
	Parsers []p.Parser
}

func (c *ParserContext) Init() *ParserContext {
	c.Parsers = []p.Parser{
		&p.WordpressParser{},
		&p.OpenAPIParser{},
		&p.GraphQLParser{},
	}

	return c
}

func (c *ParserContext) Detect(baseURL string, header http.Header, body []byte) (p.Parser, string) {
	for _, parser := range c.Parsers {
		if ok, docURL := parser.Detect(baseURL, header, body); ok {
			return parser, docURL
		}
	}
	return nil, ""
}

File: strategy.py
from typing import Optional

import httpx

from parser.parser import Parser
from parser.wordpress import WordpressParser
from parser.openapi import OpenAPIParser
from parser.graphql import GraphQLParser


class ParserContext:
    def __init__(self):
        self.parsers: list[Parser] = [
            WordpressParser(),
            OpenAPIParser(),
            GraphQLParser(),
        ]

    def Detect(self, base_url: str, header: httpx.Headers, body: str) -> tuple[Optional[Parser], str]:
        for parser in self.parsers:
            ok, doc_url = parser.Detect(base_url, header, body)
            if ok:
                return parser, doc_url
        return None, ""
